File: java/src/main/java/org/voidland/concurrent/queue/MostlyNonBlockingPortionQueue.java
package org.voidland.concurrent.queue;

import java.util.Collection;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;


public class MostlyNonBlockingPortionQueue<E>
        implements MPMC_PortionQueue<E>
{
    private AtomicInteger size;
    private int maxSize;
    
    private Object notFullCondition;
    private Object notEmptyCondition;
    private Object emptyCondition;
    
    private AtomicBoolean aProducerIsWaiting;
    private AtomicBoolean aConsumerIsWaiting;
    
    private NonBlockingQueue<E> nonBlockingQueue;
    private boolean workDone;
    
    
    public static <E> MostlyNonBlockingPortionQueue<E> createConcurrentBlownQueue(int maxSize)
    {
    	NonBlockingQueue<E> nonBlockingQueue = new ConcurrentNonBlockingQueue<E>(maxSize);
        return new MostlyNonBlockingPortionQueue<E>(maxSize, nonBlockingQueue);
    }
    
    
    public MostlyNonBlockingPortionQueue(int maxSize, NonBlockingQueue<E> nonBlockingQueue)
    {
    	this.size = new AtomicInteger(0);
        this.maxSize = maxSize;
        
        this.notFullCondition = new Object();
        this.notEmptyCondition = new Object();
        this.emptyCondition = new Object();

        this.aProducerIsWaiting = new AtomicBoolean(false);
        this.aConsumerIsWaiting = new AtomicBoolean(false);
        
        this.nonBlockingQueue = nonBlockingQueue;
        this.workDone = false;
    }
    
    @Override
    public void addPortion(E portion)
    {
        try
        {
            while (true)
            {
                if (this.size.get() >= this.maxSize)
                {
                    synchronized (this.notFullCondition)
                    {
                        while (this.size.get() >= this.maxSize)
                        {
                        	this.aProducerIsWaiting.set(true);
                            this.notFullCondition.wait();
                        }
                    }
                }
                
                if (this.nonBlockingQueue.tryEnqueue(portion))
                {
                	break;
                }
            }
            this.size.getAndIncrement();
    
            if (this.aConsumerIsWaiting.compareAndSet(true, false))
            {
                synchronized (this.notEmptyCondition)
                {
                    this.notEmptyCondition.notifyAll();
                }
            }
        }
        catch (InterruptedException e)
        {
            throw new RuntimeException(e);
        }
    }
    
    @Override
    public E retrievePortion()
    {
        try
        {
            E portion = this.nonBlockingQueue.tryDequeue();
    
            if (portion == null)
            {
                synchronized (this.notEmptyCondition)
                {
                	while (true)
                	{
                        if (this.workDone)
                        {
                            return null;
                        }
                        
                        portion = this.nonBlockingQueue.tryDequeue();
                        if (portion != null)
                        {
                        	break;
                        }
                        
                        this.aConsumerIsWaiting.set(true);
                        this.notEmptyCondition.wait();
                	}
                }
            }
            
            int newSize = this.size.decrementAndGet();
            
            if (this.aProducerIsWaiting.compareAndSet(true, false))
            {
                synchronized (this.notFullCondition)
                {
                    this.notFullCondition.notifyAll();
                }
            }
            
            if (newSize == 0)
            {
                synchronized (this.emptyCondition)
                {
                    this.emptyCondition.notify();
                }
            }
    
            return portion;
        }
        catch (InterruptedException e)
        {
            throw new RuntimeException(e);
        }
    }
    
    @Override
    public void ensureAllPortionsAreRetrieved()
    {
        try
        {
            synchronized (this.notEmptyCondition)
            {
                this.notEmptyCondition.notifyAll();
            }
    
            synchronized (this.emptyCondition)
            {
                while (this.size.get() > 0)
                {
                    this.emptyCondition.wait();
                }
            }
        }
        catch (InterruptedException e)
        {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void stopConsumers(Collection<Thread> finalConsumerThreads)
    {
        synchronized (this.notEmptyCondition)
        {
        	this.workDone = true;
            this.notEmptyCondition.notifyAll();
        }
    }
    
    @Override
    public int getSize()
    {
        return this.size.get();
    }
    
    @Override
    public int getMaxSize()
    {
        return this.maxSize;
    }
}

File: rust/src/org/voidland/concurrent/queue/mostly_non_blocking_portion_queue.rs
use super::MPMC_PortionQueue;
use super::NonBlockingQueue;
use super::non_blocking_queue::ConcurrentPortionQueue;
use super::non_blocking_queue::AsyncMPMC_PortionQueue;

use std::sync::atomic::AtomicUsize;
use std::sync::atomic::AtomicBool;
use std::sync::atomic::Ordering;
use std::sync::Mutex;
use std::sync::Condvar;
use crossbeam::utils::CachePadded;
use std::sync::MutexGuard;


struct Size
{
    size: AtomicUsize,
    maxSize: usize
}

pub struct MostlyNonBlockingPortionQueue<E>
{
    sizes: CachePadded<Size>,
    
    mutex: CachePadded<Mutex<bool>>,
    notFullCondition: CachePadded<Condvar>,
    notEmptyCondition: CachePadded<Condvar>,
    emptyCondition: CachePadded<Condvar>,
    
    aProducerIsWaiting: CachePadded<AtomicBool>,
    aConsumerIsWaiting: CachePadded<AtomicBool>,
    
    nonBlockingQueue: CachePadded<Box<dyn NonBlockingQueue<E>>>
}

impl<E: Send + Sync + 'static> MostlyNonBlockingPortionQueue<E>
{
    pub fn createConcurrentBlownQueue(maxSize: usize) -> Self
    {
        let nonBlockingQueue: Box<dyn NonBlockingQueue<E>> = Box::new(ConcurrentPortionQueue::<E>::new(maxSize));
        MostlyNonBlockingPortionQueue::new(maxSize, nonBlockingQueue)
    }
    
    pub fn createAsynMPMC_BlownQueue(maxSize: usize) -> Self
    {
        let nonBlockingQueue: Box<dyn NonBlockingQueue<E>> = Box::new(AsyncMPMC_PortionQueue::<E>::new(maxSize));
        MostlyNonBlockingPortionQueue::new(maxSize, nonBlockingQueue)
    }

    
    pub fn new(maxSize: usize, nonBlockingQueue: Box<dyn NonBlockingQueue<E>>) -> Self
    {
        Self
        {
            sizes: CachePadded::new(
                Size
                {
                    size: AtomicUsize::new(0),
                    maxSize: maxSize
                }),
            mutex: CachePadded::new(Mutex::new(false)),
            notFullCondition: CachePadded::new(Condvar::new()),
            notEmptyCondition: CachePadded::new(Condvar::new()),
            emptyCondition: CachePadded::new(Condvar::new()),
            aProducerIsWaiting: CachePadded::new(AtomicBool::new(false)),
            aConsumerIsWaiting: CachePadded::new(AtomicBool::new(false)),
            nonBlockingQueue: CachePadded::new(nonBlockingQueue)
        }
    }
    
    
    fn lockMutexIfNecessary<'a>(&'a self, lock: &mut Option<MutexGuard<'a, bool>>)
    {
        if lock.is_none()
        {
            let _ = lock.insert(self.mutex.lock().unwrap());
        }
    }
    
    fn notifyAllWaitingProducers<'a>(&'a self, lock: &mut Option<MutexGuard<'a, bool>>)
    {
        if self.aProducerIsWaiting.compare_exchange(true, false, Ordering::AcqRel, Ordering::Relaxed).is_ok()
        {
            self.lockMutexIfNecessary(lock);
            self.notFullCondition.notify_all();
        }        
    }
    
    fn notifyAllWaitingConsumers<'a>(&'a self, lock: &mut Option<MutexGuard<'a, bool>>)
    {
        if self.aConsumerIsWaiting.compare_exchange(true, false, Ordering::AcqRel, Ordering::Relaxed).is_ok()
        {
            self.lockMutexIfNecessary(lock);
            self.notEmptyCondition.notify_all();
        }        
    }

        fn waitOnCondition(cond: &Condvar, lock: &mut Option<MutexGuard<'_, bool>>)
    {
        let theLock : Option<MutexGuard<'_, bool>> = lock.take();
        let _ = lock.insert(cond.wait(theLock.unwrap()).unwrap());
    }
}

impl<E: Send + Sync + 'static> MPMC_PortionQueue<E> for MostlyNonBlockingPortionQueue<E>
{
    fn addPortion(&self, mut portion: E)
    {
        let mut lock: Option<MutexGuard<'_, bool>> = None;
        
        loop
        {
            if self.sizes.size.load(Ordering::Acquire) >= self.sizes.maxSize
            {
                self.lockMutexIfNecessary(&mut lock);
                
                loop
                {
                    let theOnlyWaitingProducer: bool = self.aProducerIsWaiting.compare_exchange(false, true, Ordering::AcqRel, Ordering::Relaxed).is_ok();
                    
                    if self.sizes.size.load(Ordering::Acquire) < self.sizes.maxSize
                    {
                        if theOnlyWaitingProducer
                        {
                            self.aProducerIsWaiting.store(false, Ordering::Release);
                        }
                        break;
                    }
                    
                    MostlyNonBlockingPortionQueue::<E>::waitOnCondition(&self.notFullCondition, &mut lock);
                }
            }
            
            match self.nonBlockingQueue.tryEnqueue(portion)
            {
                Ok(_) => break,
                Err(error) => portion = error
            }
        }
        self.sizes.size.fetch_add(1, Ordering::Release);
        
        self.notifyAllWaitingConsumers(&mut lock);
    }
    
    fn retrievePortion(&self) -> Option<E>
    {
        let mut lock: Option<MutexGuard<'_, bool>> = None;
        
        let mut portion: Option<E> = self.nonBlockingQueue.tryDequeue();
        if portion.is_none()
        {
            self.lockMutexIfNecessary(&mut lock);
            
            loop
            {
                let theOnlyWaitingConsumer: bool = self.aConsumerIsWaiting.compare_exchange(false, true, Ordering::AcqRel, Ordering::Relaxed).is_ok();
                
                portion = self.nonBlockingQueue.tryDequeue();
                if portion.is_some()
                {
                    if theOnlyWaitingConsumer
                    {
                        self.aConsumerIsWaiting.store(false, Ordering::Release);
                    }
                    break;
                }

                let workDone: bool = **lock.as_ref().unwrap();
                if workDone
                {
                    return None;
                }
                
                MostlyNonBlockingPortionQueue::<E>::waitOnCondition(&self.notEmptyCondition, &mut lock);
            }
        }
        
        let newSize: usize = self.sizes.size.fetch_sub(1, Ordering::AcqRel) - 1;
        
        self.notifyAllWaitingProducers(&mut lock);
        
        if newSize == 0
        {
            self.lockMutexIfNecessary(&mut lock);
            self.emptyCondition.notify_one();
        }
        
        portion
    }
    
    fn ensureAllPortionsAreRetrieved(&self)
    {
        let mut lock = self.mutex.lock().unwrap();
        self.notEmptyCondition.notify_all();
        while self.sizes.size.load(Ordering::Acquire) > 0
        {
            lock = self.emptyCondition.wait(lock).unwrap();
        }
    }
    
    fn stopConsumers(&self, _finalConsumerCount: usize)
    {
        let mut workDone = self.mutex.lock().unwrap();
        *workDone = true;
        self.notEmptyCondition.notify_all();
    }
    
    fn getSize(&self) -> usize
    {
        self.sizes.size.load(Ordering::Relaxed)
    }
    
    fn getMaxSize(&self) -> usize
    {
        self.sizes.maxSize
    }
}
